### ISSUE TITLE
9. Ensemble_RF_Hyperparameter

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_python-II#46]
"""
Random Forest Hyper parameter
"""

from sklearn.ensemble import RandomForestClassifier # model
from sklearn.datasets import load_wine  # dataset
from sklearn.metrics import accuracy_score # 모델 평가

### 1. dataset load
wine = load_wine()
wine.feature_names # X변수명
wine.target_names # y범주 이름 # array(['class_0', 'class_1', 'class_2'], dtype='<U7')

X = wine.data
y = wine.target

X.shape # (178, 13)

### 2. RF model
rf = RandomForestClassifier() # 객체 생성
'''
n_estimators : integer, optional (default=100) : 트리 수
criterion : string, optional (default="gini") or "entropy" : 중요변수 선정기준
max_depth : integer or None, optional (default=None) : 트리 깊이
min_samples_split : int, float, optional (default=2) : 노드 분할할때 사용되는 최소 샘플 수
min_samples_leaf : int, float, optional (default=1) : 단노드 분할 최소 샘플 수
max_features : int, float, string or None, optional (default="auto") : 최대 X변수 사용 수
n_jobs 
random_state
'''

model = rf.fit(X=X, y=y)

### 3. Grid Search
- pipeline -> grid search -> model turning
from sklearn.model_selection import GridSearchCV
for문 안쓰고 그리드 서치 할수 있는 라이브러리

- 1) params 설정
-dict 형식 = {'object__C' : params_range}
params = {'n_estimators' : [100, 200, 300, 400], 
                'max_depth' : [3,6,8,10],
                'min_samples_split' : [2,3,4,5],
               'min_samples_leaf' : [1,3,5,7]} 
- 2) GridSearchCV 객체
gs = GridSearchCV(estimator = model, param_grid = params, scoring = 'accuracy', cv=5, n_jobs =-1)
-scoring = 'accuracy', cv=10 : 점수평가방식/ 10차 교차검정/ n_jobs : cpu수
-시스템 - 장치관리자 가면 cpu 갯수 볼수 있음 
--1 : 전체 cpu 사용하겠다.

model = gs.fit(X, y)

- best score
acc = model.score(X, y) # 1.0
acc

- best parameter
model.best_params_
'''
 {'max_depth': 6,
 'min_samples_leaf': 1,
 'min_samples_split': 5,
 'n_estimators': 200}
 '''
